contracts/tvl-staking/RibbonDualStakingRewards.sol

```diff
@@ -0,0 +1,359 @@
+
+  function notifyRewardAmount(uint256 reward0, uint256 reward1)
+    external
+    override
+    onlyRewardsDistribution
+    updateReward(address(0))
+  {
+    Rewards memory _rewardRate;
+    uint256 _rewardsDuration = rewardsDuration;
+    if (block.timestamp >= periodFinish) {
+      _rewardRate = Rewards(
+        reward0.div(_rewardsDuration).toUint128(),
+        reward1.div(_rewardsDuration).toUint128()
+      );
+    } else {
+      uint256 remaining = periodFinish.sub(block.timestamp);
+      _rewardRate = rewardRate;
+      _rewardRate = Rewards(
+        reward0
+          .add(remaining.mul(_rewardRate.token0))
+          .div(_rewardsDuration)
+          .toUint128(),
+        reward1
+          .add(remaining.mul(_rewardRate.token1))
+          .div(_rewardsDuration)
+          .toUint128()
+      );
+    }
+
+    // Ensure the provided reward amount is not more than the balance in the contract.
+    // This keeps the reward rate in the right range, preventing overflows due to
+    // very high values of rewardRate in the earned and rewardsPerToken functions;
+    // Reward + leftover must be less than 2^256 / 10^18 to avoid overflow.
+    require(
+      _rewardRate.token0 <=
+        rewardsToken0.balanceOf(address(this)).div(rewardsDuration),
+      "Provided reward0 too high"
+    );
+    require(
+      _rewardRate.token1 <=
+        rewardsToken1.balanceOf(address(this)).div(rewardsDuration),
+      "Provided reward1 too high"
+    );
+
+    rewardRate = _rewardRate;
+    periodFinish = startEmission.add(rewardsDuration);
+    lastUpdateTime = lastTimeRewardApplicable();
+    emit RewardAdded(address(rewardsToken0), reward0);
+    emit RewardAdded(address(rewardsToken1), reward1);
+  }
+
+  // Added to support recovering LP Rewards from other systems such as BAL to be distributed to holders
+  function recoverERC20(address tokenAddress, uint256 tokenAmount)
+    external
+    onlyOwner
+  {
+    require(
+      tokenAddress != address(stakingToken),
+      "Cannot withdraw the staking token"
+    );
+    IERC20(tokenAddress).safeTransfer(owner, tokenAmount);
+    emit Recovered(tokenAddress, tokenAmount);
+  }
+
+  function setRewardsDuration(uint256 _rewardsDuration) external onlyOwner {
+    require(
+      block.timestamp > periodFinish,
+      "Previous rewards period must be complete before changing the duration for the new period"
+    );
+    rewardsDuration = _rewardsDuration;
+    emit RewardsDurationUpdated(rewardsDuration);
+  }
+
+  function setStartEmission(uint256 _startEmission) external onlyOwner {
+    require(
+      block.timestamp < _startEmission,
+      "Start emission must be in the future"
+    );
+    startEmission = _startEmission;
+    emit StartEmissionUpdated(startEmission);
+  }
+
+  /* ========== MODIFIERS ========== */
+
+  modifier updateReward(address account) {
+    (uint256 rewardPerToken0, uint256 rewardPerToken1) = rewardPerToken();
+    Rewards memory _rewardPerTokenStored = Rewards(
```

Comment:
you cant set a storage ref to a memory. have to do like this:

1. remove _rewardPerTokenStored
2. `rewardPerTokenStored = Rewards(rewardPerToken0.toUint128(), rewardPerToken1.toUint128())`

Reply:
`_rewardPerTokenStored` allows the Rewards struct to only be initialized once (also means the safecasts don't need to be repeated)
reinitializing adds quite a bit of gas on each call